### ISSUE TITLE
Failed to open script file '/Main.as'

### ISSUE THREAD

**Issue by lgbaldoni** [yvt/openspades#947]
I'm packaging 0.1.3 for openSUSE and I'm seeing that on startup.
It's probably a path error on my part, but I couldn't locate that file anywhere.

Any hints?

**Comment by yvt** [comment]
`Scripts/Main.as` is provided by `Resources/pak002-Base.pak`, which should be installed to the directory at path [`${CMAKE_INSTALL_PREFIX}/${OPENSPADES_INSTALL_RESOURCES}`](https://github.com/yvt/openspades/blob/2d972c00c7ae63aaa58108039c876113e22b9c10/Sources/Gui/Main.cpp#L397-L398), which defaults to `/usr/local/share/games/openspades/Resources`. (Pak files are actually zip files, so you can check their contents by `zipinfo`, etc.)

OpenSpades lists all search paths when launching. E.g., on macOS, it displays the following:

```
$ openspades
2017/12/15 05:43:17 [Main.cpp:324] Package: OpenSpades 0.1.1
2017/12/15 05:43:17 [DirectoryFileSystem.cpp:46] Directory File System Initialized: ./Resources (Read-only)
2017/12/15 05:43:17 [DirectoryFileSystem.cpp:46] Directory File System Initialized: /usr/local/share/games/openspades/Resources (Read-only)
2017/12/15 05:43:17 [Main.cpp:400] XDG_DATA_HOME is /home/REDACTED
2017/12/15 05:43:17 [DirectoryFileSystem.cpp:46] Directory File System Initialized: /home/REDACTED/.local/share/openspades/Resources (Read/Write)
...
```

**Comment by lgbaldoni** [comment]
Sorry, I forgot to attach the [logfile](https://github.com/yvt/openspades/files/6249344/SystemMessages.log): pak002-Base.pak appears to be loaded and it contains Main.as.

**Comment by lgbaldoni** [comment]
On second thought, it looks like there's a hardcoded build path:|

**Comment by lgbaldoni** [comment]
Ok, I give up. I have no idea how to pinpoint the source of this problem. Any hint?

**Comment by yvt** [comment]
It appears that the error message `Failed to open script file '/Main.as'` originates from the implementation of `CScriptBuilder::LoadScriptSection`:

https://github.com/yvt/openspades/blob/dcf449c845d509bbde9d79239b32d201c1389961/Sources/AngelScript/addons/scriptbuilder.cpp#L169-L173

However, this message is not supposed to show up because this method is overridden by our class `ScriptBuilder`:

https://github.com/yvt/openspades/blob/dcf449c845d509bbde9d79239b32d201c1389961/Sources/ScriptBindings/ScriptManager.cpp#L48-L55

So, the overriding did not work for some reason. Maybe try `make clean all`? Also setting a breakpoint on `CScriptBuilder::LoadScriptSection` and checking the backtrace at that point might tell us why this supposedly unused code was called.

**Comment by lgbaldoni** [comment]
Will try all that as soon as possible. Meanwhile, could you tell me how the unpacking mechanism works? I mean, how does openspades extract Main.as from the .pak file and where does it put it?

**Comment by lgbaldoni** [comment]
I hope [this](https://github.com/yvt/openspades/files/6252537/openspades_gdb.txt) is what you meant.

I forgot to specify that I'm using actionscript libraries built separately: do you think this might be part of the problem?

**Comment by yvt** [comment]
> I forgot to specify that I'm using angelscript libraries built separately: do you think this might be part of the problem?

Yes, that explains why the overriding didn't work; `CScriptBuilder::LoadScriptSection` isn't `virtual` in the original version of AngelScript.

**Comment by lgbaldoni** [comment]
I've tested with your fork and that error has disappeared.
Do you see a way around that? Like upstreaming the patch or something? Shipping static libraries with packages is often met with grumbles :unamused: 

Now I'm seeing [this](https://github.com/yvt/openspades/files/6252614/openspades_log_2.txt) on startup, but I assume that's due to limitations of the VM driver.

But let me bother you about a further matter: is openal used at all? It's mentioned in a CMakeLists.txt and source files, but it doesn't appear to be linked.

Regards

**Comment by yvt** [comment]
> I've tested with your fork and that error has disappeared.
> Do you see a way around that? Like upstreaming the patch or something? Shipping static libraries with packages is often met with grumbles 😒

No, I haven't considered that; I'm not familiar with their contribution process.

> Now I'm seeing [this](https://github.com/yvt/openspades/files/6252614/openspades_log_2.txt) on startup, but I assume that's due to limitations of the VM driver.

#895

> But let me bother you about a further matter: is openal used at all? It's mentioned in a CMakeLists.txt and source files, but it doesn't appear to be linked.

It's loaded at runtime by `dlopen` when the game is started (i.e., when you press the start button). The path is specified by [`s_alDriver` config variable](https://github.com/yvt/openspades/blob/46679c65c4d8794de3f3c4bc78ad30be5e82d6de/Sources/Audio/ALFuncs.cpp#L32).
